Firestore/core/test/firebase/firestore/local/remote_document_cache_test.mm

```diff
@@ -0,0 +1,222 @@
+        };
+        OptionalMaybeDocumentMap read = cache_->GetAll(DocumentKeySet{
+            testutil::Key(kDocPath),
+            testutil::Key(kLongDocPath),
+            testutil::Key("foo/nonexistent"),
+        });
+        EXPECT_THAT(read, HasAtLeastDocs(written));
+        auto found = read.find(DocumentKey::FromPathString("foo/nonexistent"));
+        ASSERT_TRUE(found != read.end());
+        ASSERT_EQ(absl::nullopt, found->second);
+      });
+}
+
+TEST_P(RemoteDocumentCacheTest, SetAndReadADocumentAtDeepPath) {
+  SetAndReadTestDocument(kLongDocPath);
+}
+
+TEST_P(RemoteDocumentCacheTest, SetAndReadDeletedDocument) {
+  persistence_->Run("test_set_and_read_deleted_document", [&] {
+    absl::optional<MaybeDocument> deleted_doc = DeletedDoc(kDocPath, kVersion);
+    cache_->Add(*deleted_doc);
+
+    ASSERT_EQ(cache_->Get(testutil::Key(kDocPath)), deleted_doc);
+  });
+}
+
+TEST_P(RemoteDocumentCacheTest, SetDocumentToNewValue) {
+  persistence_->Run("test_set_document_to_new_value", [&] {
+    SetTestDocument(kDocPath);
+    absl::optional<MaybeDocument> new_doc =
+        Doc(kDocPath, kVersion, Map("data", 2));
+    cache_->Add(*new_doc);
+    ASSERT_EQ(cache_->Get(testutil::Key(kDocPath)), new_doc);
+  });
+}
+
+TEST_P(RemoteDocumentCacheTest, RemoveDocument) {
+  persistence_->Run("test_remove_document", [&] {
+    SetTestDocument(kDocPath);
+    cache_->Remove(testutil::Key(kDocPath));
+
+    ASSERT_EQ(cache_->Get(testutil::Key(kDocPath)), absl::nullopt);
+  });
+}
+
+TEST_P(RemoteDocumentCacheTest, RemoveNonExistentDocument) {
+  persistence_->Run("test_remove_non_existent_document", [&] {
+    // no-op, but make sure it doesn't throw.
+    EXPECT_NO_THROW(cache_->Remove(testutil::Key(kDocPath)));
+  });
+}
+
+// TODO(mikelehen): Write more elaborate tests once we have more elaborate
+// implementations.
+TEST_P(RemoteDocumentCacheTest, DocumentsMatchingQuery) {
+  persistence_->Run("test_documents_matching_query", [&] {
+    // TODO(rsgowman): This just verifies that we do a prefix scan against the
+    // query path. We'll need more tests once we add index support.
+    SetTestDocument("a/1");
+    SetTestDocument("b/1");
+    SetTestDocument("b/1/z/1");
+    SetTestDocument("b/2");
+    SetTestDocument("c/1");
+
+    core::Query query = Query("b");
+    DocumentMap results = cache_->GetMatching(query);
+    std::vector<Document> docs = {
+        Doc("b/1", kVersion, kDocData),
+        Doc("b/2", kVersion, kDocData),
+    };
+    EXPECT_THAT(results.underlying_map(), HasExactlyDocs(docs));
+    //    ExpectMapHasDocs(results.underlying_map(), docs, /* exactly= */ true);
+  });
+}
+
+// MARK: - Helpers
+
+Document RemoteDocumentCacheTest::SetTestDocument(
+    const absl::string_view path) {
+  Document doc = Doc(path, kVersion, kDocData);
+  cache_->Add(doc);
+  return doc;
+}
+
+void RemoteDocumentCacheTest::SetAndReadTestDocument(
+    const absl::string_view path) {
+  persistence_->Run("SetAndReadTestDocument", [&] {
```

Comment:
This method is only used once and could potentially be inlined.

Reply:
It's used by both `SetAndReadADocument` and `SetAndReadADocumentAtDeepPath`.